File: Ahorn/entities/cassetteField.jl
module CherryHelperCassetteFieldModule

using ..Ahorn, Maple

@mapdef Entity "CherryHelper/CassetteField" CassetteField(x::Integer, y::Integer, width::Integer=8, height::Integer=8, index::Integer=0, tempo::Number=1.0,Type::String="ZipMover")
const colorsA=[(0.0, 0.53, 1.0, 0.4),(1.0, 0.23, 0.79, 0.4),(1.0, 0.79, 0.0, 0.4),(0.54, 1.0, 0.22, 0.4)]
const colorsB=[(0.0, 0.53, 1.0, 1.0),(1.0, 0.23, 0.79, 1.0),(1.0, 0.79, 0.0, 1.0),(0.54, 1.0, 0.22, 1.0)]

Ahorn.minimumSize(entity::CassetteField) = 8, 8
Ahorn.resizable(entity::CassetteField) = true, true

Ahorn.selection(entity::CassetteField) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CassetteField, room::Maple.Room)
    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))
	index = Int(get(entity.data, "index",0))
    Ahorn.drawRectangle(ctx, 0, 0, width, height, colorsA[index+1], colorsB[index+1])
end

const colorNames = Dict{String, Int}(
    "Blue" => 0,
    "Rose" => 1,
    "Bright Sun" => 2,
    "Malachite" => 3
)

const placements = Ahorn.PlacementDict(
    "Cassette Field ($index - $color, Cherry Helper)" => Ahorn.EntityPlacement(
        CassetteField,
        "rectangle",
        Dict{String, Any}(
            "index" => index,
        )
    ) for (color, index) in colorNames
)

Ahorn.editingOptions(entity::CassetteField) = Dict{String, Any}(
    "index" => colorNames
)


Ahorn.minimumSize(entity::CassetteField) = 16, 16
Ahorn.resizable(entity::CassetteField) = true, true

Ahorn.selection(entity::CassetteField) = Ahorn.getEntityRectangle(entity)

end

File: Loenn/entities/cassetteField.lua
local cassetteField = {}

local colors = {
    Blue = 0,
    Rose = 1,
    Bright_Sun = 2,
    Malachite = 3
}

cassetteField.name = "CherryHelper/CassetteField"
cassetteField.placements = {}

for color, indx in pairs(colors) do
    table.insert(cassetteField.placements, {
        name = color,
        width = 16,
        height = 16,
        index = indx,
        tempo = 1.0,
        Type = "ZipMover",
        placementType = "rectangle"
    })
end

function cassetteField.fillColor(room, entity)
    if entity.index == 0 then
        return {0.0, 0.53, 1.0, 0.4}
    elseif entity.index == 1 then
        return {1.0, 0.23, 0.79, 0.4}
    elseif entity.index == 2 then
        return {1.0, 0.79, 0.0, 0.4}
    elseif entity.index == 3 then
        return {0.54, 1.0, 0.22, 0.4}
    end
end

function cassetteField.borderColor(room, entity)
    if entity.index == 0 then
        return {0.0, 0.53, 1.0}
    elseif entity.index == 1 then
        return {1.0, 0.23, 0.79}
    elseif entity.index == 2 then
        return {1.0, 0.79, 0.0}
    elseif entity.index == 3 then
        return {0.54, 1.0, 0.22}
    end
end

return cassetteField